### PR TITLE
Authorization service abstraction

### PR DESCRIPTION
**Pull request by cgilbert76**
Kept the existing `IAuthService` interface, and renamed the `AuthService` concrete implementation to `GoogleFitAuthService`.  The methods as declared in `IAuthService` should serve as the minimum needed for any platform that needs to be supported.  For any exceptions to this, a new platform's auth service can utilize an interface which inherits from `IAuthService`, and the needed methods/properties added there.

Within the implementations of the `IAuthService` methods, a specific platform's endpoints can be used in concrete types that implement the `IAuthRequest<TResponseType>` interface.  For example, `GoogleFitAuthUriRequest` implements `IGoogleFitAuthUriRequest`, which implements `IAuthRequest<AuthUriResponse>`.  `ExecuteAsync` within `GoogleFitAuthUriRequest` looks the same as what `AuthUriRequest` `ExecuteAsync` did, but now it is in a platform specific class.

To summarize, when adding a new authorization service you would:

1. Create an interface which inherits from `IAuthService`, adding any platform specific exceptions as necessary
2. Create the concrete type which corresponds to this interface.
3. Create interfaces for the platform's authorization endpoint requests, that inherit from `IAuthRequest<TResponseType>`.
4. Create the concrete types for the endpoint request interfaces.
5. Use the injected interfaces of those types in the new auth service within the appropriate methods specified by the interface used in step 1, to enable the auth flow for the platform.

Replaced the helper method `GetAuthFlow` that was in all of the various AuthRequest classes with a Singleton declaration in the DI container.

### CODE REVIEW

File: src/GoogleFitOnFhir/Services/GoogleFitAuthService.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+namespace GoogleFitOnFhir.Services
+{
+    public class GoogleFitAuthService : IGoogleFitAuthService
+    {
+        private readonly ILogger<GoogleFitAuthService> _logger;
+        private readonly GoogleFitClientContext _clientContext;
+        private readonly IGoogleFitAuthUriRequest _googleFitAuthUriRequest;
+        private readonly IGoogleFitAuthTokensRequest _googleFitAuthTokensRequest;
+        private readonly IGoogleFitRefreshTokenRequest _googleFitRefreshTokensRequest;
```
Comment (namalu) [inline src/GoogleFitOnFhir/Services/GoogleFitAuthService.cs:21]: I am starting to feel like these 3 classes (GoogleFitAuthTokensRequest, GoogleFitAuthUriRequest, and GoogleFitRefreshTokensRequest are adding an unneeded layer go abstraction that is causing code bloat and making testing difficult.

This IGoogleFitAuthService already serves as a layer of abstraction for the Google endpoints. I think we can move away from these request classes and simply "new up" the instance of GoogleAuthorizationCodeFlow in the constructor of this class, make the 3 calls represented by the request classes directly on the GoogleAuthorizationCodeFlow object and convert the responses to the correct return types.

This would mean that this class is not testable, but I'm fine with that because this is just a wrapper for the Google APIS.

Comment (namalu) [inline src/GoogleFitOnFhir/Services/GoogleFitAuthService.cs:21]: I am starting to feel like these 3 classes (GoogleFitAuthTokensRequest, GoogleFitAuthUriRequest, and GoogleFitRefreshTokensRequest are adding an unneeded layer go abstraction that is causing code bloat and making testing difficult.

This IGoogleFitAuthService already serves as a layer of abstraction for the Google endpoints. I think we can move away from these request classes and simply "new up" the instance of GoogleAuthorizationCodeFlow in the constructor of this class, make the 3 calls represented by the request classes directly on the GoogleAuthorizationCodeFlow object and convert the responses to the correct return types.

This would mean that this class is not testable, but I'm fine with that because this is just a wrapper for the Google APIS.

File: src/GoogleFitOnFhir.Identity/Startup.cs [hunk 2]
```diff
@@ -35,12 +39,38 @@ public override void Configure(IFunctionsHostBuilder builder)
             builder.Services.AddSingleton(sp => new GoogleFitClientContext(googleFitClientId, googleFitClientSecret, hostName));
             builder.Services.AddSingleton(sp => new StorageAccountContext(storageAccountConnectionString));
             builder.Services.AddSingleton(sp => new SecretClient(new Uri(usersKeyVaultUri), new DefaultAzureCredential()));
+            builder.Services.AddSingleton(sp =>
```
Comment (namalu) [inline src/GoogleFitOnFhir.Identity/Startup.cs:42]: I would like to avoid this pattern as much as possible. Rather than "manually" resolving dependencies, I would prefer to let the DI framework do all the work. See my comment about removing the 3 request classes. I think we can just pass the GoogleFitClientContext to the constructor of the GoogleFitAuthService.

I also think that the Scopes array should be a property on the GoogleFitClientContext named DefaultScopes. At some point, we will need to support dynamically adding additional scopes to the request, but we will probably always need the https://www.googleapis.com/auth/userinfo.profile scope at the very least, and it should be the default scope.

Comment (namalu) [inline src/GoogleFitOnFhir.Identity/Startup.cs:42]: I would like to avoid this pattern as much as possible. Rather than "manually" resolving dependencies, I would prefer to let the DI framework do all the work. See my comment about removing the 3 request classes. I think we can just pass the GoogleFitClientContext to the constructor of the GoogleFitAuthService.

I also think that the Scopes array should be a property on the GoogleFitClientContext named DefaultScopes. At some point, we will need to support dynamically adding additional scopes to the request, but we will probably always need the https://www.googleapis.com/auth/userinfo.profile scope at the very least, and it should be the default scope.

File: src/GoogleFitOnFhir/Clients/GoogleFit/GoogleFitClientContext.cs [hunk 3]
```diff
@@ -20,12 +22,24 @@ public GoogleFitClientContext(string clientId, string clientSecret, string hostN
                 .Append("/api/googlefit/callback");
 
             CallbackUri = stringBuilder.ToString();
+
+            DefaultScopes = new[]
```
Comment (namalu) [inline src/GoogleFitOnFhir/Clients/GoogleFit/GoogleFitClientContext.cs:26]: Let's add a comment here that these aren't the final default values.

File: src/GoogleFitOnFhir/Services/IAuthRequest.cs [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace GoogleFitOnFhir.Services
+{
+    public interface IAuthRequest<TResponseType>
```
Comment (namalu) [inline src/GoogleFitOnFhir/Services/IAuthRequest.cs:11]: Looks like this Interface can be deleted.

**Review by namalu — APPROVED** [review]
